lib/rgeo/impl_helper/valid_op.rb

```diff
@@ -0,0 +1,357 @@
+# frozen_string_literal: true
+
+module RGeo
+  module ImplHelper
+    module TopologyErrors
+      # Standard error messages from
+      # https://github.com/locationtech/jts/blob/0afbfb1956ec24912a8b4dc4edff0f1200442857/modules/core/src/main/java/org/locationtech/jts/operation/valid/TopologyValidationError.java#L98-L110
+      TOPOLOGY_VALIDATION_ERR = "Topology Validation Error"
+      REPEATED_POINT = "Repeated Point"
+      HOLE_OUTSIDE_SHELL = "Hole lies outside shell"
+      NESTED_HOLES = "Holes are nested"
+      DISCONNECTED_INTERIOR = "Interior is disconnected"
+      SELF_INTERSECTION = "Self-intersection"
+      RING_SELF_INTERSECTION = "Ring Self-intersection"
+      NESTED_SHELLS = "Nested shells"
+      DUPLICATE_RINGS = "Duplicate Rings"
+      TOO_FEW_POINTS = "Too few distinct points in geometry component"
+      INVALID_COORDINATE = "Invalid Coordinate"
+      UNCLOSED_RING = "Ring is not closed"
+    end
+
+    # Mixin based off of the JTS/GEOS IsValidOp class.
+    # Implements #valid? and #invalid_reason on Features that include this.
+    #
+    # @see https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/operation/valid/IsValidOp.java
+    module ValidOp
+      # TODO: determine if "Unkown Validity" should count as valid.
+      # Validity of geometry
+      #
+      # @return Boolean
+      def valid?
+        invalid_reason.nil?
+      end
+      alias is_valid? valid?
+
+      # Reason for invalidity or nil if valid
+      #
+      # @return String
+      def invalid_reason
+        return @invalid_reason if defined?(@invalid_reason)
+        @invalid_reason = check_valid
+      end
```

Comment:
In my branch I do not memoize the reason, but rather have an `invalid_reason_memo` that does it. Anyway, I think we can leave that there for now, and discuss it in my PR when we'll have the complete flow!